grub-btrfs: update to 4.8.1 + add runit service

FollieHiyuki opened the pull request:
<!-- Mark items with [x] where applicable -->

#### General
- [ ] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [ ] I generally don't use the affected packages but briefly tested this PR

<!--
If GitHub CI cannot be used to validate the build result (for example, if the
build is likely to take several hours), make sure to
[skip CI](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#continuous-integration).
When skipping CI, uncomment and fill out the following section.
Note: for builds that are likely to complete in less than 2 hours, it is not
acceptable to skip CI.
-->
<!-- 
#### Does it build and run successfully? 
(Please choose at least one native build and, if supported, at least one cross build. More are better.)
- [ ] I built this PR locally for my native architecture, (ARCH-LIBC)
- [ ] I built this PR locally for these architectures (if supported. mark crossbuilds):
  - [ ] aarch64-musl
  - [ ] armv7l
  - [ ] armv6l-musl
-->

FollieHiyuki commented:
This is just a minor version, but I want to include a replacement for upstream's systemd service if possible.
@ericonr can you make some suggestions about this?

ericonr commented:
I'd suggest something close to how we have `dracut-uefi`. It's a sub package which pulls in the dependencies specific to UEFI, and also includes some files of our own (in this case, the kernel hooks; in your case, it'd be the service file, assuming it can be implemented cleanly).

FollieHiyuki commented:
I will look into it. For now please don't merge

FollieHiyuki commented:
I can't think of any ways to make the service file without additional dependencies.
My implementation with `inotify-tools` is kind of messy, so I shamelessly took the `wendy` route from @circhioz.
Advice is welcomed.

ericonr commented:
Please mention in the commit message that you added the service :)

ericonr commented on srcpkgs/grub-btrfs/files/grub-btrfs/run:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+[ -r conf ] && . ./conf
+exec wendy -m 768 -w "${SNAPSHOTS_PATH}" bash -c \
+ 'if [ -s "/boot/grub/grub-btrfs.cfg" ]; then /etc/grub.d/41_snapshots-btrfs; else update-grub; fi' 2>&1
```
```suggestion
[ -r conf ] && . ./conf
exec 2>&1
exec wendy -m 768 -w "${SNAPSHOTS_PATH}" bash -c \
 'if [ -s "/boot/grub/grub-btrfs.cfg" ]; then /etc/grub.d/41_snapshots-btrfs; else update-grub; fi'
```

ericonr reviewed (commented):
I think this probably warrants adding a logger, can you please add one? Grep for `vlogger` to see how it's done.

@ahesford since I know he uses wendy.

FollieHiyuki commented:
I added the commit message. Will look into `vlogger` later

FollieHiyuki commented:
I added a `-v` option to wendy, since it might be useful when logging.
The log contains the output of `update-grub` too, which doesn't seem useful to me. Should we disable logging for the command output (only display when there are errors)?

ericonr commented:
Honestly, your choice. I don't use grub or btrfs. Make a service you're comfortable with, we can improve some piece of it if someone suggests changes or complains.

FollieHiyuki commented:
I guess we can keep the service as it is right now. Seeing some more log doesn't hurt.

ahesford commented on srcpkgs/grub-btrfs/files/grub-btrfs/run:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+[ -r conf ] && . ./conf
+exec 2>&1
```
You probably want to put this above the conf source just in case there are some error messages printed while sourcing the conf file.

ahesford commented on srcpkgs/grub-btrfs/files/grub-btrfs/run:
```diff
@@ -0,0 +1,6 @@
+
+[ -r conf ] && . ./conf
+exec 2>&1
+exec wendy -v -m 768 -w "${SNAPSHOTS_PATH}" bash -c \
```
1. Omit the `-v` here and add some logic that will recognize a variable like `VERBOSE` that can be set in the `conf` file to govern the addition of `-v`. This could be a catch-all `OPTS` variable like other services uses, but there probably isn't much value in allowing other flags, since you want tighter control on the path and mode anyway.
2. Is mode 768 sufficient? If snapshots can be renamed, you might also need bits for `IN_MOVE` or the `IN_MOVED_{TO,FROM}`.  You should do some experiments to see what kind of events a snapshot create/rename/destroy trigger.
3. You should probably verify that `SNAPSHOTS_PATH` is non-empty and either fail or substitute a default when it is.

ahesford reviewed (commented):
I know nothing about btrfs and next to nothing about GRUB, so I can't weigh in on the intended functionality of this service, but as a matter of using `wendy`, I left a few generic suggestions for improvement. I wouldn't worry about the `wendy` dependency; using raw `inotify-tools` in scripts gets ugly fast, especially if you want to properly forward signals to the service.

FollieHiyuki commented:
Thanks @ahesford for all the feedbacks.
The only features of btrfs I use are compression and snapshoting, so I doubt I know any better :)
> Omit the -v here and add some logic that will recognize a variable like VERBOSE that can be set in the conf file to govern the addition of -v. This could be a catch-all OPTS variable like other services uses, but there probably isn't much value in allowing other flags, since you want tighter control on the path and mode anyway.

I keep the OPTS variable. There could be a use case for it, for example `-r` when the user manages snapshots inside multiple sub directories. Adding `-d` would be a sanity check too.
> Is mode 768 sufficient? If snapshots can be renamed, you might also need bits for IN_MOVE or the IN_MOVED_{TO,FROM}. You should do some experiments to see what kind of events a snapshot create/rename/destroy trigger.

I added IN_MOVE bit.
AFAIK btrfs snapshots are subvolumes, and they can be treated as normal directories (delete wirh `rm -r`, rename wirh `mv`) so I guess it emits events as normal directories in the file system. Apparently the service at least works on my system with all 3 events.
> You should probably verify that SNAPSHOTS_PATH is non-empty and either fail or substitute a default when it is.

Since `wendy` will do nothing if the directory doesn't exist, I added the check for that case. Having to check for SNAPSHOTS_PATH value as well will be complicating, so I keep a default value in the `conf` file

ericonr commented:
Is this ready to be reviewed?

FollieHiyuki commented:
Yes please do

ericonr commented on srcpkgs/grub-btrfs/files/grub-btrfs/run:
```diff
@@ -0,0 +1,12 @@
+
+if [ -d "${SNAPSHOTS_PATH}" ]
+then
+	exec wendy ${OPTS} -m 960 -w "${SNAPSHOTS_PATH}" bash -c \
```
I don't think `bash` should be hardcoded here, but it's guaranteed to be pulled in by the package, so leaving as is for now.

FollieHiyuki replied:
@ericonr that command is taken from upstream's systemd service. I don't think they will switch to POSIX sh, so it would be fine

ahesford replied:
Upstream might use bash in a systemd service, but they aren't providing a runit service, so we should adopt Void idioms when adapting our own. There is nothing specific to bash, so /bin/sh is the way to go here.

FollieHiyuki replied:
That sounds reasonable. I will make the change
